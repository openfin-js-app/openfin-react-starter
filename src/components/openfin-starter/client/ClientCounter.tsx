import * as React from 'react';
import cx from 'classnames';
import Button from '@material-ui/core/Button';

import { WithStyles, withStyles} from '@material-ui/core/styles';


interface IProps {
    count:number,
    onChange:(count:number)=>void
}

const ClientCounterComp:React.FunctionComponent<IProps> = (
    {
        count,onChange,
    }
)=>{

    const handleIncBtnClick = ()=>{
        onChange(
            count+1
        )
    }

    const handleDecBtnClick = ()=>{
        onChange(
            count-1
        )
    }

    return(<React.Fragment>
        <span>{count}</span>
        <Button variant="contained" color="primary"
                onClick={handleIncBtnClick}
        >
            Increase
        </Button>
        <Button variant="contained" color="secondary"
                onClick={handleDecBtnClick}
        >
            Decrease
        </Button>
    </React.Fragment>)

}

export default ClientCounterComp;