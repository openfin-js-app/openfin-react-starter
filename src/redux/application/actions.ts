import { createAction, ActionFunctionAny, Action } from 'redux-actions';
import { IApplicationNewSnackbarOption, IApplicationSetSnackbarStatusOption, IApplicationCloseSnackBarOption } from './types';

export const APPLICATION_STARTED:string = 'APPLICATION_STARTED';
export const APPLICATION_READY:string = 'APPLICATION_READY';
export const APPLICATION_DRAWER_TOGGLE:string = 'APPLICATION_DRAWER_TOGGLE';
export const APPLICATION_TOGGLE_WINDOW_STATE:string = 'APPLICATION_TOGGLE_WINDOW_STATE';

export const applicationStarted:ActionFunctionAny<Action<void>> = createAction(APPLICATION_STARTED);
export const applicationReady:ActionFunctionAny<Action<void>> = createAction(APPLICATION_READY);
export const applicationDrawerToggle:ActionFunctionAny<Action<void>> = createAction(APPLICATION_DRAWER_TOGGLE);
export const applicationToogleWindowState:ActionFunctionAny<Action<void>> = createAction(APPLICATION_TOGGLE_WINDOW_STATE);

// snackbar

export const APPLICATION_NEW_SNACKBAR='APPLICATION_NEW_SNACKBAR';
export const APPLICATION_SET_SNACKBAR_STATUS='APPLICATION_SET_SNACKBAR_STATUS';
export const APPLICATION_PROCESS_SNACKBAR_QUEUE='APPLICATION_PROCESS_SNACKBAR_QUEUE';
export const APPLICATION_CLOSE_SNACKBAR='APPLICATION_CLOSE_SNACKBAR';

export const applicationNewSnackbar:ActionFunctionAny<Action<IApplicationNewSnackbarOption>>
    = createAction(APPLICATION_NEW_SNACKBAR, (option:IApplicationNewSnackbarOption)=>(option));
export const applicationSetSnackbarStatus:ActionFunctionAny<Action<IApplicationSetSnackbarStatusOption>>
    = createAction(APPLICATION_SET_SNACKBAR_STATUS, (option:IApplicationSetSnackbarStatusOption)=>(option));
export const applicationProcessSnackbarQueue:ActionFunctionAny<Action<void>>
    = createAction(APPLICATION_PROCESS_SNACKBAR_QUEUE);
export const applicationCloseSnackbar:ActionFunctionAny<Action<IApplicationCloseSnackBarOption>>
    = createAction(APPLICATION_CLOSE_SNACKBAR, (option)=>(option));


// launch bar

export const APPLICATION_LAUNCH_BAR_TOGGLE = 'APPLICATION_LAUNCH_BAR_TOGGLE';
export const APPLICATION_LAUNCH_BAR_TOGGLE_COLLAPSE = 'APPLICATION_LAUNCH_BAR_TOGGLE_COLLAPSE';
export const APPLICATION_LAUNCH_NEW_WINDOW = 'APPLICATION_LAUNCH_NEW_WINDOW';
export const applicationLaunchBarToggle:ActionFunctionAny<Action<void>> = createAction(APPLICATION_LAUNCH_BAR_TOGGLE);
export const applicationLaunchBarToggleCollapse:ActionFunctionAny<Action<void>> = createAction(APPLICATION_LAUNCH_BAR_TOGGLE_COLLAPSE);
export const applicationLaunchNewWindow:ActionFunctionAny<Action<any>>
    = createAction(APPLICATION_LAUNCH_NEW_WINDOW,(appJson:any)=>(appJson));