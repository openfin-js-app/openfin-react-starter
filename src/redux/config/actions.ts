import {createAction, ActionFunctionAny, Action} from 'redux-actions';

export const CONFIG_RESET:string = 'CONFIG_RESET';
export const CONFIG_UPDATE_ONE_FIELD:string = 'CONFIG_UPDATE_ONE_FIELD';

export const configReset = createAction(CONFIG_RESET, (option)=>(option));
export const configUpdateOneField = createAction(CONFIG_UPDATE_ONE_FIELD, (option)=>(option));

// optional actions depending on the config application values

export const CONFIG_UPDATE_NEW_WINDOW_POSITION:string = 'CONFIG_UPDATE_NEW_WINDOW_POSITION';
export const configUpdateNewWindowPosition = createAction(CONFIG_UPDATE_NEW_WINDOW_POSITION);