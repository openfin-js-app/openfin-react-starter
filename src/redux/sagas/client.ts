import { buffers, delay } from 'redux-saga';
import { all, call, put, take, takeLatest, takeEvery, fork, select, actionChannel } from 'redux-saga/effects';
import { SHARED_ACTION_ORIGIN_TAG } from '@albertli90/redux-openfin/channel';

import {
    CLIENT_SET_VALUE,
    applicationNewSnackbar,
} from '..';

export function* handleTakingClientSetValue(action) {
    const payload:any = action.payload;

    if (payload[SHARED_ACTION_ORIGIN_TAG] !== window[SHARED_ACTION_ORIGIN_TAG]){
        yield put(applicationNewSnackbar({
            message:`Received msg from ${payload[SHARED_ACTION_ORIGIN_TAG]}`,
            variant:'info'
        }))
    }else{
        yield put(applicationNewSnackbar({
            message:`Message sent at ${payload[SHARED_ACTION_ORIGIN_TAG]}`,
            variant:'success'
        }))
    }

}

export default function *() {
    yield takeEvery(CLIENT_SET_VALUE,handleTakingClientSetValue);
}