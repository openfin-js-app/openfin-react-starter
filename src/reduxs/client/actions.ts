import { createAction, ActionFunctionAny, Action } from 'redux-actions';
import { IClientState } from './types';

export const CLIENT_SET_VALUE:string = 'CLIENT_SET_VALUE';
export const clientSetValue:ActionFunctionAny<Action<Partial<IClientState>>> = createAction(CLIENT_SET_VALUE,
    (option:Partial<IClientState>)=>(option)
);