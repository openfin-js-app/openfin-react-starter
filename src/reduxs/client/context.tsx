import * as React from 'react';

import { Omit } from 'react-openfin/utils/typeHelper';

import { IClientState } from './types'

export interface IWithClientActions{
    updateClientCount:(count:number)=>void
}

export interface IWithClient {
    state:IClientState,
    actions:IWithClientActions,
}

export const ClientContext = React.createContext<Partial<IWithClient>|null>(null);

const { Provider, Consumer } = ClientContext;

interface IWithClientContext {
    clientContext?:Partial<IWithClient>
}

export type WithClientContext = IWithClientContext;

export const ClientContextProvider = Provider;
export const ClientContextConsumer = Consumer;


export const withClientContext:<
    P extends WithClientContext
    >(Component: React.ComponentType<P>)=>React.FunctionComponent<Omit<P,WithClientContext>> =

    <P extends WithClientContext>(Component: React.ComponentType<P>)=> {
        return function ComponentWithConfig(props:Omit<P,WithClientContext>){
            return (
                <ClientContextConsumer>
                    {value =>
                        // @ts-ignore
                        (<Component {...props} clientContext={value}/>)}
                </ClientContextConsumer>
            );}
    };
