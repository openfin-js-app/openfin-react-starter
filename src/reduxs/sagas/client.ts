import { delay, put, putResolve, takeEvery, takeLatest } from 'redux-saga/effects';
import { SHARED_ACTION_ORIGIN_TAG } from 'redux-openfin/channel';

// !!!README!!!
// use the redux for advanced features
import {
    applicationNewSnackbar,
    APPLICATION_AWAIT,
    APPLICATION_CHILD_AWAIT,
    APPLICATION_NOTIFICATION_AWAIT,
    applicationReady,
    applicationChildReady,
    applicationNotificationReady,
    APPLICATION_CUR_WIN_CLOSING,
    applicationCurWinReadyToClose,
    APPLICATION_LAUNCH_BAR_TOGGLED,
    // types
    APPLICATION_LAUNCH_BAR_STATUS,
} from 'react-openfin/reduxs';

import {
    CLIENT_SET_VALUE,
} from '..';

export function* handleTakingClientSetValue(action) {
    if (action[SHARED_ACTION_ORIGIN_TAG] !== window[SHARED_ACTION_ORIGIN_TAG]){
        // !!!README!!!
        // use the redux for advanced features
        // once createReactOpenfinMiddleware is imported, redux-openfin&react-openfin's certain actions can be directly
        // triggered via client redux
        yield put(applicationNewSnackbar({
            message:`Received msg from ${action[SHARED_ACTION_ORIGIN_TAG]}`,
            variant:'info'
        }))
    }else{
        // !!!README!!!
        // use the redux for advanced features
        // once createReactOpenfinMiddleware is imported, redux-openfin&react-openfin's certain actions can be directly
        // triggered via client redux
        yield put(applicationNewSnackbar({
            message:`Message sent at ${action[SHARED_ACTION_ORIGIN_TAG]}`,
            variant:'success'
        }))
    }
}

// !!!README!!!
// trigger client side initialization over here if needed
export function* handleStarting(action){
    console.log('client saga :: handlStarting',action);
    if (action.type === APPLICATION_AWAIT){
        // APPLICATION_AWAIT will be sent once application started
        // client side initialization effects can be triggered over here
        yield delay(3000);
        // once done, client could send applicationReady action to let react-openfin to switch to
        // the targetUrl specified int the payload from loading view before fuse timeout
        yield putResolve(applicationReady({
            // optional sample targetUrl
            // targetUrl:'/login'
        }));
    }else if(action.type === APPLICATION_NOTIFICATION_AWAIT){
        // APPLICATION_NOTIFICATION_AWAIT will be sent once a notification started
        // once done, client could send applicationNotificationReady action to let react-openfin to switch to
        // the a certain targetUrl specified int the payload from original url before fuse timeout
        yield putResolve(applicationNotificationReady({
            // optional sample targetUrl
            // targetUrl:'/sampleUrl'
        }));
    }else if(action.type === APPLICATION_CHILD_AWAIT){
        // APPLICATION_CHILD_AWAIT will be sent once a child window started
        // once done, client could send applicationChildReady action to let react-openfin to switch to
        // the a certain targetUrl specified int the payload from original url before fuse timeout
        yield putResolve(applicationChildReady({
            // optional sample targetUrl
            // targetUrl:'/sampleUrl'
        }));
    }
}

export function* handleAppClosing(action){
    console.log('client saga::handleAppClosing');
    yield putResolve(applicationCurWinReadyToClose());
}

export function* handleLaunchbarToggled(action) {
    const { status } = action.payload;

    if (status === APPLICATION_LAUNCH_BAR_STATUS.SWITCH_TO_MAIN_WIN){
        yield put(applicationNewSnackbar({
            message:`Switch to Main Window`,
            variant:'primary'
        }))
    }

}

export default function *() {
    yield takeEvery(CLIENT_SET_VALUE,handleTakingClientSetValue);

    // !!!README!!!
    // trigger client side initialization over here if needed
    yield takeLatest(APPLICATION_AWAIT, handleStarting);
    yield takeLatest(APPLICATION_CHILD_AWAIT, handleStarting);
    yield takeLatest(APPLICATION_NOTIFICATION_AWAIT, handleStarting);
    yield takeLatest(APPLICATION_CUR_WIN_CLOSING, handleAppClosing);
    yield takeLatest(APPLICATION_LAUNCH_BAR_TOGGLED, handleLaunchbarToggled);
}