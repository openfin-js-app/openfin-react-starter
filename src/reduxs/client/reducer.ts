import { handleActions, Action } from 'redux-actions';

import { IClientState,  } from './types';

import {CLIENT_SET_VALUE} from './actions';

export const defaultState:Partial<IClientState>={
    count:0,
};


export default (parentWindowState?:Partial<IClientState>)=>{
    let initState:Partial<IClientState>;

    if (parentWindowState){
        initState={
            ...parentWindowState,
        }
    }else{
        initState = defaultState;
    }

    return handleActions({
        [CLIENT_SET_VALUE]:(state,action)=>{
            return {
                ...state,
                ...action.payload,
            }
        }
    },initState);

}