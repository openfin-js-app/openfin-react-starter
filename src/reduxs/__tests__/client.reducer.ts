import * as Actions from '../client/actions';

import reducerCreator, { defaultState }  from '../client/reducer';

describe('Client reducer', ()=>{
    describe('With parent state', () => {
        let reducer

        beforeAll(()=>{
            reducer = reducerCreator(defaultState);
        })

        it ('CLIENT_SET_VALUE',()=>{
            const action = Actions.clientSetValue({count:1});
            expect(reducer(undefined,action)).toMatchSnapshot();
        })
    })
    describe('Without parent state', () => {
        let reducer

        beforeAll(()=>{
            reducer = reducerCreator();
        })

        it ('CLIENT_SET_VALUE',()=>{
            const action = Actions.clientSetValue({count:1});
            expect(reducer(undefined,action)).toMatchSnapshot();
        })
    })
})