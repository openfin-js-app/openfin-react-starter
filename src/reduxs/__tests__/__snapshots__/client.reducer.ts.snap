// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client reducer With parent state CLIENT_SET_VALUE 1`] = `
Object {
  "count": 1,
}
`;

exports[`Client reducer Without parent state CLIENT_SET_VALUE 1`] = `
Object {
  "count": 1,
}
`;
